File: go/test/invasive_specie_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/colombia-public-sdk"
	"github.com/voxgig-sdk/colombia-public-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestInvasiveSpecieEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.InvasiveSpecie(nil)
		if ent == nil {
			t.Fatal("expected non-nil InvasiveSpecieEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := invasive_specieBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "invasive_specie." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set COLOMBIAPUBLIC_TEST_INVASIVE_SPECIE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		invasiveSpecieRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.invasive_specie", setup.data)))
		var invasiveSpecieRef01Data map[string]any
		if len(invasiveSpecieRef01DataRaw) > 0 {
			invasiveSpecieRef01Data = core.ToMapAny(invasiveSpecieRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = invasiveSpecieRef01Data

		// LIST
		invasiveSpecieRef01Ent := client.InvasiveSpecie(nil)
		invasiveSpecieRef01Match := map[string]any{}

		invasiveSpecieRef01ListResult, err := invasiveSpecieRef01Ent.List(invasiveSpecieRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, invasiveSpecieRef01ListOk := invasiveSpecieRef01ListResult.([]any)
		if !invasiveSpecieRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", invasiveSpecieRef01ListResult)
		}

		// LOAD
		invasiveSpecieRef01MatchDt0 := map[string]any{
			"id": invasiveSpecieRef01Data["id"],
		}
		invasiveSpecieRef01DataDt0Loaded, err := invasiveSpecieRef01Ent.Load(invasiveSpecieRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		invasiveSpecieRef01DataDt0LoadResult := core.ToMapAny(invasiveSpecieRef01DataDt0Loaded)
		if invasiveSpecieRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if invasiveSpecieRef01DataDt0LoadResult["id"] != invasiveSpecieRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func invasive_specieBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "invasive_specie", "InvasiveSpecieTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read invasive_specie test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse invasive_specie test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"invasive_specie01", "invasive_specie02", "invasive_specie03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("COLOMBIAPUBLIC_TEST_INVASIVE_SPECIE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"COLOMBIAPUBLIC_TEST_INVASIVE_SPECIE_ENTID": idmap,
		"COLOMBIAPUBLIC_TEST_LIVE":      "FALSE",
		"COLOMBIAPUBLIC_TEST_EXPLAIN":   "FALSE",
		"COLOMBIAPUBLIC_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["COLOMBIAPUBLIC_TEST_INVASIVE_SPECIE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["COLOMBIAPUBLIC_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["COLOMBIAPUBLIC_APIKEY"],
			},
			extra,
		})
		client = sdk.NewColombiaPublicSDK(core.ToMapAny(mergedOpts))
	}

	live := env["COLOMBIAPUBLIC_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["COLOMBIAPUBLIC_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/InvasiveSpecieEntityTest.php
<?php
declare(strict_types=1);

// InvasiveSpecie entity test

require_once __DIR__ . '/../colombiapublic_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class InvasiveSpecieEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = ColombiaPublicSDK::test(null, null);
        $ent = $testsdk->InvasiveSpecie(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = invasive_specie_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "invasive_specie." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set COLOMBIAPUBLIC_TEST_INVASIVE_SPECIE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $invasive_specie_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.invasive_specie")));
        $invasive_specie_ref01_data = null;
        if (count($invasive_specie_ref01_data_raw) > 0) {
            $invasive_specie_ref01_data = Helpers::to_map($invasive_specie_ref01_data_raw[0][1]);
        }

        // LIST
        $invasive_specie_ref01_ent = $client->InvasiveSpecie(null);
        $invasive_specie_ref01_match = [];

        [$invasive_specie_ref01_list_result, $err] = $invasive_specie_ref01_ent->list($invasive_specie_ref01_match, null);
        $this->assertNull($err);
        $this->assertIsArray($invasive_specie_ref01_list_result);

        // LOAD
        $invasive_specie_ref01_match_dt0 = [
            "id" => $invasive_specie_ref01_data["id"],
        ];
        [$invasive_specie_ref01_data_dt0_loaded, $err] = $invasive_specie_ref01_ent->load($invasive_specie_ref01_match_dt0, null);
        $this->assertNull($err);
        $invasive_specie_ref01_data_dt0_load_result = Helpers::to_map($invasive_specie_ref01_data_dt0_loaded);
        $this->assertNotNull($invasive_specie_ref01_data_dt0_load_result);
        $this->assertEquals($invasive_specie_ref01_data_dt0_load_result["id"], $invasive_specie_ref01_data["id"]);

    }
}

function invasive_specie_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/invasive_specie/InvasiveSpecieTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = ColombiaPublicSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["invasive_specie01", "invasive_specie02", "invasive_specie03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("COLOMBIAPUBLIC_TEST_INVASIVE_SPECIE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "COLOMBIAPUBLIC_TEST_INVASIVE_SPECIE_ENTID" => $idmap,
        "COLOMBIAPUBLIC_TEST_LIVE" => "FALSE",
        "COLOMBIAPUBLIC_TEST_EXPLAIN" => "FALSE",
        "COLOMBIAPUBLIC_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["COLOMBIAPUBLIC_TEST_INVASIVE_SPECIE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["COLOMBIAPUBLIC_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["COLOMBIAPUBLIC_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new ColombiaPublicSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["COLOMBIAPUBLIC_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["COLOMBIAPUBLIC_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
